Fixed out of range core to bank id mappings for L1 small memory region allocations

mpiseTT opened the pull request:
[Max Pool 2D test doesn't work in slow dispatch mode](https://github.com/tenstorrent/tt-metal/pull/19949#top)
#16191 

### Ticket
16191

### Problem description
This change fixes a bank id out of range assert and a data verification error in maxpool2d tests. The problem was that for L1 small allocations the core to bank id mapping was different from L1 allocations, when it seems like they should be the same (since they are both partitions in the same physical memory).

### What's changed
Gave the same core to bank id mappings as L1 for L1 small region
Also added tt_metal/api gtests for sharded L1 small region memory write then read back verification

### Checklist
- [X] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes: https://github.com/tenstorrent/tt-metal/actions/runs/14180541154
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [X] New/Existing tests provide coverage for changes

nhuang-tt commented on tests/tt_metal/tt_metal/api/test_sharded_l1_buffer.cpp:
```diff
@@ -45,19 +47,19 @@ namespace local_test_functions {
-bool l1_buffer_read_write(IDevice* device, const L1Config& test_config) {
-    bool pass = true;
-
+auto l1_buffer_write_wait(IDevice* device, const L1Config& test_config) {
```
I'm not a big fan of using `auto` return type for functions as it hides the return type

mpiseTT replied:
Yes, I can understand that, and especially here because the return type is quite complex (which contradictorily fits the use case for using `auto`!). I don't hold a strong position on this, so if no one else wants to chime in, I'll make change it.

abhullar-tt commented on tests/tt_metal/tt_metal/common/device_fixture.hpp:
```diff
@@ -47,15 +47,22 @@ class DeviceFixture : public DispatchFixture {
     }
 
+    DeviceFixture() = default;
+    DeviceFixture(size_t l1_small_size) : DispatchFixture(l1_small_size) {}
```
rather than just l1_small_size, could you make this accept some `TestDeviceFixtureConfig` so we could toggle some of the other defaults easier

abhullar-tt replied:
perhaps just the parameters that DevicePool is constructed with (currently they are set to some  defaults)

mpiseTT replied:
There is only one other, `DEFAULT_TRACE_REGION_SIZE`. The other arguments to `DevicePool::initialize` are a std::vector of device ids, and # of devices and core config. The vector would be inconvenient to initialize with a default value, because it's a vector. For # of devices, not sure if it is a good idea to have a default; in DispatchFixture class it is a dynamic value (from `RunTimeOptions`). The last one is core config, which is also dynamic from `RunTimeOptions`. So, how about if I just add `DEFAULT_TRACE_REGION_SIZE` as another configurable parameter in this new ctor, and leave it initialized with the default value that all the tests seem to be using today?

abhullar-tt commented on tt_metal/impl/allocator/l1_banking_allocator.cpp:
```diff
@@ -127,6 +132,11 @@ void Allocator::init_compute_and_storage_l1_bank_manager() {
             logical_core_to_bank_ids_[BufferType::L1].insert({logical_core, {remapped_bank_id}});
             bank_id_to_logical_core_.insert({remapped_bank_id, logical_core});
             bank_id_to_bank_offset.insert({remapped_bank_id, 0});
+            if (config_.l1_small_size > 0) {
```
there is a comment in this file above mentioning that l1-small banks don't need to be reshuffled. Could you update that so it mentions that l1 and l1-small both map to same bank

abhullar-tt reviewed (commented):
Nice! Could you also run BH post commit pipeline

mpiseTT commented:
> Nice! Could you also run BH post commit pipeline

I was hoping that "All post commit tests" meant all platforms. I found "Blackhole post commit tests" and I have started a run in it here: https://github.com/tenstorrent/tt-metal/actions/runs/14185659756

mpiseTT commented:
> > Nice! Could you also run BH post commit pipeline
> 
> I was hoping that "All post commit tests" meant all platforms. I found "Blackhole post commit tests" and I have started a run in it here: https://github.com/tenstorrent/tt-metal/actions/runs/14185659756

The run is completed and one test failed:
[  FAILED  ] 1 test, listed below:
[  FAILED  ] SiliconDriverBH.UnalignedStaticTLB_RW

I've triggered rerun of that sub-suite now.

mpiseTT commented:
> > > Nice! Could you also run BH post commit pipeline
> > 
> > 
> > I was hoping that "All post commit tests" meant all platforms. I found "Blackhole post commit tests" and I have started a run in it here: https://github.com/tenstorrent/tt-metal/actions/runs/14185659756
> 
> The run is completed and one test failed: [ FAILED ] 1 test, listed below: [ FAILED ] SiliconDriverBH.UnalignedStaticTLB_RW
> 
> I've triggered rerun of that sub-suite now.

This failed again, but it doesn't seem related to these changes. This time it was a startup failure:
`error parsing called workflow
".github/workflows/blackhole-post-commit.yaml"`

mpiseTT committed:
Core to bank id mapping for L1-small memory region is now same as for the L1 region

This change fixes a bank id out of range assert and a data verification error in maxpool2d tests.
Also added tt_metal/api gtests for sharded L1 small region memory write then read back verification
